brushfire-training/src/main/scala/com/stripe/brushfire/local/Trainer.scala

```diff
@@ -126,26 +126,21 @@ case class Trainer[K: Ordering, V: Ordering, T: Monoid](
     copy(trees = newTrees)
   }
 
-  def validate[P, E](error: Error[T, P, E])(implicit voter: Voter[T, P]): Option[E] = {
-    var output: Option[E] = None
-    trainingData.foreach{instance =>
-      val predictions =
-        for (
-          (treeIndex, tree) <- treeMap
-            if sampler.includeInValidationSet(instance.id, instance.timestamp, treeIndex);
-          target <- tree.targetFor(instance.features).toList
-        ) yield target
-
-      if(!predictions.isEmpty) {
-        val newError = error.create(instance.target, voter.combine(predictions))
-        output = output
-                    .map{old => error.semigroup.plus(old, newError)}
-                    .orElse(Some(newError))
-      }
-    }
-
-    output
-  }
+  def validate[P, E](error: Error[T, P, E])(implicit voter: Voter[T, P]): Option[E] =
+    error.semigroup.sumOption(trainingData.iterator.map { instance =>
+      val id = instance.id
+      val timestamp = instance.timestamp
+      val features = instance.features
+
+      val predictions = treeMap.iterator
+        .filter { case (treeIndex, _) =>
+          sampler.includeInValidationSet(id, timestamp, treeIndex)
+        }.flatMap { case (_, tree) =>
+          tree.targetFor(features)
+        }.toVector
+
+      error.create(instance.target, voter.combine(predictions))
```

Comment:
I think this is going to give you the wrong answer when `predictions` is empty (eg because this instance is not in the validation set). Who knows what `voter.combine(Vector.empty)` is going to produce, but it's totally possible (even, likely) that it will produce a non-zero error, which means we're accumulating all kinds of extra error for stuff that should have been filtered out.